pkg/vacuum/vacuum.go

```diff
@@ -0,0 +1,223 @@
+package vacuum
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	sqlStart            = "SELECT pg_try_advisory_lock(1982010619820106)"
+	sqlListChunks       = "SELECT format('%I.%I', schema_name, table_name) from _ps_catalog.chunks_to_vacuum_freeze LIMIT 1000"
+	sqlVacuumFmt        = "VACUUM (FREEZE, ANALYZE) %s"
+	sqlStop             = "SELECT pg_advisory_unlock(1982010619820106)"
+	delay               = 10 * time.Second
+	defaultDisable      = false
+	defaultRunFrequency = 10 * time.Minute
+	defaultParallelism  = 4
+	minParallelism      = 1
+)
+
+type Config struct {
+	Disable      bool
+	RunFrequency time.Duration
+	Parallelism  int
+}
+
+func ParseFlags(fs *flag.FlagSet, cfg *Config) *Config {
+	fs.BoolVar(&cfg.Disable, "vacuum.disable", defaultDisable, "disables the vacuum engine")
+	fs.DurationVar(&cfg.RunFrequency, "vacuum.run-frequency", defaultRunFrequency, "how often should the vacuum engine run")
+	fs.IntVar(&cfg.Parallelism, "vacuum.parallelism", defaultParallelism, "how many goroutines/connections should be used to vacuum")
+	return cfg
+}
+
+func Validate(cfg *Config) error {
+	if cfg.Disable {
+		return nil
+	}
+	if cfg.Parallelism < minParallelism {
+		return fmt.Errorf("vacuum.parallelism must be at least %d: %d", minParallelism, cfg.Parallelism)
+	}
+	if cfg.RunFrequency <= 0 {
+		return fmt.Errorf("vacuum.run-frequency must be positive: %d", cfg.RunFrequency)
+	}
+	return nil
+}
+
+// Engine periodically vacuums compressed chunks
+type Engine struct {
+	runFreq     time.Duration
+	pool        pgxconn.PgxConn
+	parallelism int
+	mu          sync.Mutex
+	kill        func()
+}
+
+// NewEngine creates a new Engine
+func NewEngine(pool pgxconn.PgxConn, runFreq time.Duration, parallelism int) *Engine {
+	return &Engine{
+		runFreq:     runFreq,
+		pool:        pool,
+		parallelism: parallelism,
+	}
+}
+
+// Start starts the Engine
+// Blocks forever unless Stop is called
+func (e *Engine) Start() {
+	var execute, kill = every(e.runFreq, e.RunOnce)
+	func() {
+		e.mu.Lock()
+		defer e.mu.Unlock()
+		e.kill = kill
+	}()
+	execute() // blocks forever unless kill is called
+}
+
+// Stop stops the engine if it is running
+func (e *Engine) Stop() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.kill != nil {
+		e.kill()
+	}
+}
+
+// every executes a task periodically
+// returns an execute function which when called will block
+// and execute task periodically, and a kill function which will
+// terminate the execution function if called
+func every(every time.Duration, task func()) (execute, kill func()) {
+	var ticker = time.NewTicker(every)
+	var once sync.Once
+	var die = make(chan struct{}, 1)
+
+	kill = func() {
+		once.Do(func() {
+			ticker.Stop()
+			die <- struct{}{}
```

Comment:
you shouldn't need to send anything, just close. `After calling close, and after any previously sent values have been received, receive operations will return the zero value for the channel's type without blocking.`

Reply:
https://github.com/timescale/promscale/pull/1625/commits/ec4cd25a431cba7e06bd082650d5b43738aa3c20